File: KG_Assist/src/process.c
/**
 * KG Assist - 进程操作模块
 * 功能: 进程枚举、句柄管理、模块遍历
 *
 * 技术要点:
 * - CreateToolhelp32Snapshot 枚举进程
 * - OpenProcess 打开目标进程
 * - Module32FirstW/NextW 遍历模块 (宽字符版本)
 * - 句柄安全管理
 */

#include "../include/common.h"

/**
 * 宽字符转 ANSI 字符串 (前置声明)
 */
static VOID WidCharToMbs(const wchar_t* wide, char* ansi, u32 ansiSize);

/* ============================================================
 * 进程枚举
 * ============================================================ */

/**
 * 枚举所有运行中的进程
 * @param processes 输出缓冲区
 * @param maxCount 最大进程数
 * @return 找到的进程数量
 */
s32 KgEnumAllProcesses(KgProcessInfo* processes, u32 maxCount) {
    if (!processes || maxCount == 0) return 0;
    
    HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
    if (snapshot == INVALID_HANDLE_VALUE) {
        KG_ERROR("创建进程快照失败");
        return 0;
    }
    
    PROCESSENTRY32W entry = {0};
    entry.dwSize = sizeof(entry);
    
    s32 count = 0;
    
    if (Process32FirstW(snapshot, &entry)) {
        do {
            if (count >= (s32)maxCount) break;
            
            // 填充进程信息
            memset(&processes[count], 0, sizeof(KgProcessInfo));
            processes[count].pid = entry.th32ProcessID;
            processes[count].parentPid = entry.th32ParentProcessID;
            
            // 拷贝进程名 (宽字符, 直接拷贝)
            if (entry.szExeFile[0]) {
                wcscpy_s(processes[count].name, MAX_PATH, entry.szExeFile);
            }
            
            count++;
        } while (Process32NextW(snapshot, &entry));
    }
    
    CloseHandle(snapshot);
    KG_DEBUG("枚举到 %d 个进程", count);
    return count;
}

/**
 * 按名称查找进程
 * @param name 进程名 (大小写不敏感, ANSI 或宽字符)
 * @param info 输出进程信息
 * @return 是否找到
 */
BOOL KgFindProcess(const char* name, KgProcessInfo* info) {
    if (!name || !info) return FALSE;
    
    // 转换查找名到宽字符
    wchar_t wideName[MAX_PATH] = {0};
    MultiByteToWideChar(CP_ACP, 0, name, -1, wideName, MAX_PATH);
    
    HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
    if (snapshot == INVALID_HANDLE_VALUE) {
        KG_ERROR("创建进程快照失败");
        return FALSE;
    }
    
    PROCESSENTRY32W entry = {0};
    entry.dwSize = sizeof(entry);
    
    BOOL found = FALSE;
    
    if (Process32FirstW(snapshot, &entry)) {
        do {
            // 大小写不敏感比较 (宽字符版本)
            if (_wcsicmp(entry.szExeFile, wideName) == 0) {
                // 填充进程信息
                memset(info, 0, sizeof(KgProcessInfo));
                info->pid = entry.th32ProcessID;
                info->parentPid = entry.th32ParentProcessID;
                
                // 拷贝进程名
                if (entry.szExeFile[0]) {
                    wcscpy_s(info->name, MAX_PATH, entry.szExeFile);
                }
                
                found = TRUE;
                KG_INFO("找到进程 %s, PID: %lu", name, info->pid);
                break;
            }
        } while (Process32NextW(snapshot, &entry));
    }
    
    CloseHandle(snapshot);
    
    if (!found) {
        KG_DEBUG("未找到进程: %s", name);
    }
    
    return found;
}

/**
 * 查找多个匹配的进程 (如多个游戏实例)
 */
s32 KgFindAllProcesses(const char* name, KgProcessInfo* results, u32 maxResults) {
    if (!name || !results || maxResults == 0) return 0;
    
    wchar_t wideName[MAX_PATH] = {0};
    MultiByteToWideChar(CP_ACP, 0, name, -1, wideName, MAX_PATH);
    
    HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
    if (snapshot == INVALID_HANDLE_VALUE) return 0;
    
    PROCESSENTRY32W entry = {0};
    entry.dwSize = sizeof(entry);
    
    s32 count = 0;
    
    if (Process32FirstW(snapshot, &entry)) {
        do {
            if (count >= (s32)maxResults) break;
            
            if (_wcsicmp(entry.szExeFile, wideName) == 0) {
                memset(&results[count], 0, sizeof(KgProcessInfo));
                results[count].pid = entry.th32ProcessID;
                results[count].parentPid = entry.th32ParentProcessID;
                
                if (entry.szExeFile[0]) {
                    wcscpy_s(results[count].name, MAX_PATH, entry.szExeFile);
                }
                
                count++;
            }
        } while (Process32NextW(snapshot, &entry));
    }
    
    CloseHandle(snapshot);
    return count;
}

/* ============================================================
 * 进程句柄管理
 * ============================================================ */

/**
 * 打开进程获取句柄
 * @param info 进程信息 (需要 pid)
 * @param accessRights 访问权限
 * @return 是否成功
 */
BOOL KgOpenProcess(KgProcessInfo* info, DWORD accessRights) {
    if (!info || info->pid == 0) return FALSE;
    
    // 如果已有句柄, 先关闭
    if (info->handle != NULL) {
        CloseHandle(info->handle);
        info->handle = NULL;
    }
    
    info->handle = OpenProcess(accessRights, FALSE, info->pid);
    if (info->handle == NULL) {
        // 尝试较低权限
        info->handle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ,
                                     FALSE, info->pid);
    }
    
    if (info->handle != NULL) {
        info->accessRights = accessRights;
        KG_INFO("打开进程 PID %lu 成功 (权限: 0x%08lX)", info->pid, accessRights);
        return TRUE;
    }
    
    KG_ERROR("打开进程 PID %lu 失败 (错误: %lu)", info->pid, GetLastError());
    return FALSE;
}

/**
 * 关闭进程句柄
 */
VOID KgCloseProcess(KgProcessInfo* info) {
    if (info && info->handle != NULL) {
        CloseHandle(info->handle);
        info->handle = NULL;
        info->accessRights = 0;
        KG_DEBUG("关闭进程句柄 PID: %lu", info->pid);
    }
}

/* ============================================================
 * 模块枚举
 * ============================================================ */

/**
 * 枚举进程加载的所有模块
 * @param info 进程信息 (需要有效的句柄)
 * @return 找到的模块数量
 */
BOOL KgEnumModules(KgProcessInfo* info) {
    if (!info || info->handle == NULL) return FALSE;
    
    HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPMODULE | TH32CS_SNAPMODULE32,
                                                info->pid);
    if (snapshot == INVALID_HANDLE_VALUE) {
        KG_ERROR("创建模块快照失败 (PID: %lu)", info->pid);
        return FALSE;
    }
    
    MODULEENTRY32W entry = {0};
    entry.dwSize = sizeof(entry);
    
    info->moduleCount = 0;
    
    if (Module32FirstW(snapshot, &entry)) {
        do {
            if (info->moduleCount >= KG_MAX_MODULES) break;
            
            // 填充模块信息
            KgModuleInfo* mod = &info->modules[info->moduleCount];
            memset(mod, 0, sizeof(KgModuleInfo));
            
            // 模块名 (宽字符 -> ANSI)
            WidCharToMbs(entry.szModule, mod->name, KG_MAX_PATH);
            
            // 基址和大小
            mod->baseAddress = (u32)(ULONG_PTR)entry.hModule;
            mod->sizeOfImage = entry.modBaseSize;
            
            info->moduleCount++;
        } while (Module32NextW(snapshot, &entry));
    }
    
    CloseHandle(snapshot);
    KG_DEBUG("枚举到 %u 个模块 (PID: %lu)", info->moduleCount, info->pid);
    return TRUE;
}

/**
 * 宽字符转 ANSI 字符串
 */
static VOID WidCharToMbs(const wchar_t* wide, char* ansi, u32 ansiSize) {
    if (!wide || !ansi || ansiSize == 0) return;
    WideCharToMultiByte(CP_ACP, 0, wide, -1, ansi, ansiSize, NULL, NULL);
}

/**
 * 查找指定模块
 * @param info 进程信息
 * @param name 模块名 (支持通配符前缀匹配)
 * @return 模块指针 (失败返回 NULL)
 */
KgModuleInfo* KgFindModule(KgProcessInfo* info, const char* name) {
    if (!info || !name) return NULL;
    
    for (u32 i = 0; i < info->moduleCount; i++) {
        // 精确匹配 (大小写不敏感)
        if (_stricmp(info->modules[i].name, name) == 0) {
            return &info->modules[i];
        }
        
        // 前缀匹配 (支持 "kernel32*" 这样的模式)
        size_t len = strlen(name);
        if (len > 0 && name[len - 1] == '*') {
            if (_strnicmp(info->modules[i].name, name, len - 1) == 0) {
                return &info->modules[i];
            }
        }
    }
    
    KG_DEBUG("未找到模块: %s", name);
    return NULL;
}

/**
 * 获取主模块信息
 */
KgModuleInfo* KgGetMainModule(KgProcessInfo* info) {
    if (!info || info->moduleCount == 0) return NULL;
    
    // 主模块通常是第一个
    return &info->modules[0];
}

/* ============================================================
 * 辅助函数
 * ============================================================ */

/**
 * 打印进程详细信息 (调试用)
 */
VOID KgDumpProcessInfo(const KgProcessInfo* info) {
    if (!info) return;

    KG_INFO("=== 进程信息 ===");
    KG_INFO("  PID: %lu  父PID: %lu",
            (unsigned long)info->pid,
            (unsigned long)info->parentPid);
    KG_INFO("  句柄: %p  权限: 0x%08lX  模块数: %u",
            info->handle,
            (unsigned long)info->accessRights,
            info->moduleCount);

    for (u32 i = 0; i < info->moduleCount && i < 10; i++) {
        KG_INFO("  [%u] %s  基址: 0x%08X  大小: 0x%08X",
                i, info->modules[i].name,
                info->modules[i].baseAddress,
                info->modules[i].sizeOfImage);
    }
    if (info->moduleCount > 10) {
        KG_INFO("  ... 还有 %u 个模块", info->moduleCount - 10);
    }
}

File: KG_Assist/rust/src/process.rs
//! 进程操作
//!
//! 对应 KG.exe sub_4115c0 (进程操作处理):
//!   - OpenProcess + Process32First/Next 枚举
//!   - Module32First/Next 模块枚举
//!   - 读取 PE 头获取 SizeOfImage

use crate::native_api::*;
use windows_sys::Win32::{
    Foundation::{CloseHandle, INVALID_HANDLE_VALUE},
    System::Diagnostics::ToolHelp::{
        CreateToolhelp32Snapshot, Process32FirstW, Process32NextW,
        Module32FirstW, Module32NextW,
        PROCESSENTRY32W, MODULEENTRY32W, TH32CS_SNAPPROCESS, TH32CS_SNAPMODULE,
        TH32CS_SNAPMODULE32,
    },
    System::Threading::OpenProcess,
};

pub const PROCESS_ALL_ACCESS: u32 = 0x1F0FFF;

#[derive(Clone)]
pub struct ProcessInfo {
    pub pid: u32,
    pub parent_pid: u32,
    pub name: String,
    pub handle: HANDLE,
    pub modules: Vec<ModuleInfo>,
}

#[derive(Clone)]
pub struct ModuleInfo {
    pub name: String,
    pub base: usize,
    pub size: u32,
    pub entry_point: u32,
}

/// 查找进程 — 对应 KG 的进程搜索逻辑
pub fn find_process(name: &str) -> Option<ProcessInfo> {
    unsafe {
        let snap = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
        if snap == INVALID_HANDLE_VALUE as _ {
            return None;
        }

        let mut entry: PROCESSENTRY32W = core::mem::zeroed();
        entry.dwSize = core::mem::size_of::<PROCESSENTRY32W>() as u32;

        if Process32FirstW(snap, &mut entry) != 0 {
            loop {
                let pname = utf16_to_string(entry.szExeFile.as_ptr());
                if pname.to_lowercase() == name.to_lowercase() {
                    CloseHandle(snap);
                    return Some(ProcessInfo {
                        pid: entry.th32ProcessID,
                        parent_pid: entry.th32ParentProcessID,
                        name: pname,
                        handle: 0,
                        modules: Vec::new(),
                    });
                }
                if Process32NextW(snap, &mut entry) == 0 {
                    break;
                }
            }
        }

        CloseHandle(snap);
        None
    }
}

/// 打开进程
pub fn open_process(info: &mut ProcessInfo, access: u32) -> bool {
    unsafe {
        let h = OpenProcess(access, 0, info.pid);
        if h == 0 {
            return false;
        }
        info.handle = h as HANDLE;
        true
    }
}

/// 枚举模块 — 对应 KG 的模块枚举逻辑
pub fn enum_modules(info: &ProcessInfo) -> Vec<ModuleInfo> {
    let mut modules = Vec::new();

    unsafe {
        let snap = CreateToolhelp32Snapshot(
            TH32CS_SNAPMODULE | TH32CS_SNAPMODULE32,
            info.pid,
        );
        if snap == INVALID_HANDLE_VALUE as _ {
            return modules;
        }

        let mut entry: MODULEENTRY32W = core::mem::zeroed();
        entry.dwSize = core::mem::size_of::<MODULEENTRY32W>() as u32;

        if Module32FirstW(snap, &mut entry) != 0 {
            loop {
                let mname = utf16_to_string(entry.szModule.as_ptr());
                modules.push(ModuleInfo {
                    name: mname,
                    base: entry.modBaseAddr as usize,
                    size: entry.modBaseSize as u32,
                    entry_point: 0, // 从 PE 头读取
                });
                if Module32NextW(snap, &mut entry) == 0 {
                    break;
                }
            }
        }

        CloseHandle(snap);
    }

    modules
}

/// 获取主模块 (第一个, 通常是 exe 本身)
pub fn get_main_module(info: &ProcessInfo) -> Option<&ModuleInfo> {
    info.modules.first()
}

/// 读取远程 PE 头获取 SizeOfImage 和 EntryPoint
pub fn read_pe_info(h_process: HANDLE, base: usize) -> (u32, u32) {
    // 读 DOS header (64 字节)
    let mut dos_header = [0u8; 64];
    if !remote_read(h_process, base as LPVOID, &mut dos_header) {
        return (0, 0);
    }

    // e_lfanew at offset 0x3C
    let e_lfanew = u32::from_le_bytes([
        dos_header[0x3C], dos_header[0x3D],
        dos_header[0x3E], dos_header[0x3F],
    ]) as usize;

    // 读 PE header (从 e_lfanew 开始, 248 字节足够)
    let mut pe_header = [0u8; 248];
    if !remote_read(h_process, (base + e_lfanew) as LPVOID, &mut pe_header) {
        return (0, 0);
    }

    // Signature at +0 (4 bytes, "PE\0\0")
    // FileHeader at +4 (20 bytes)
    // OptionalHeader at +24
    //   Magic at +24 (2 bytes, 0x10B = PE32)
    //   AddressOfEntryPoint at +24+16 (4 bytes)
    //   SizeOfImage at +24+56 (4 bytes)
    let entry_point = u32::from_le_bytes([
        pe_header[24 + 16], pe_header[24 + 17],
        pe_header[24 + 18], pe_header[24 + 19],
    ]);
    let size_of_image = u32::from_le_bytes([
        pe_header[24 + 56], pe_header[24 + 57],
        pe_header[24 + 58], pe_header[24 + 59],
    ]);

    (size_of_image, entry_point)
}

/// 读取 PE 时间戳 (TimeDateStamp)
pub fn read_pe_timestamp(h_process: HANDLE, base: usize) -> u32 {
    let mut dos_header = [0u8; 64];
    if !remote_read(h_process, base as LPVOID, &mut dos_header) {
        return 0;
    }
    let e_lfanew = u32::from_le_bytes([
        dos_header[0x3C], dos_header[0x3D],
        dos_header[0x3E], dos_header[0x3F],
    ]) as usize;

    // TimeDateStamp at PE+4 (FileHeader offset 4)
    let mut buf = [0u8; 8];
    if !remote_read(h_process, (base + e_lfanew) as LPVOID, &mut buf) {
        return 0;
    }
    u32::from_le_bytes([buf[4], buf[5], buf[6], buf[7]])
}

/// 关闭进程句柄
pub fn close_process(info: &mut ProcessInfo) {
    if info.handle != 0 {
        unsafe { CloseHandle(info.handle as _); }
        info.handle = 0;
    }
}

/// UTF-16 字符串转 Rust String
fn utf16_to_string(ptr: *const u16) -> String {
    let mut len = 0;
    unsafe {
        while *ptr.add(len) != 0 {
            len += 1;
        }
        let slice = core::slice::from_raw_parts(ptr, len);
        String::from_utf16_lossy(slice)
    }
}

/// 查找 LeagueClient 安装路径
///
/// 对应 KG FUN_00401b4e + FUN_00402616:
///   FUN_00401b4e = 取 LeagueClient.exe 进程模块路径
///   FUN_00402616 = 提取目录部分 (去掉文件名)
///
/// 查找顺序:
///   1. 先找正在运行的 LeagueClient.exe 进程, 取进程路径
///   2. 没找到则检查常见安装目录
pub fn find_league_client_install_path() -> Option<String> {
    // [1] 取正在运行的 LeagueClient.exe 路径
    if let Some(mut pi) = find_process("LeagueClient.exe") {
        if open_process(&mut pi, 0x0400 | 0x0010) { // PROCESS_QUERY_LIMITED_INFORMATION | PROCESS_VM_READ
            unsafe {
                extern "system" {
                    fn QueryFullProcessImageNameW(
                        hProcess: usize,
                        dwFlags: u32,
                        lpExeName: *mut u16,
                        lpdwSize: *mut u32,
                    ) -> i32;
                }
                let mut buf: [u16; 520] = [0; 520];
                let mut size: u32 = 520;
                let ok = QueryFullProcessImageNameW(
                    pi.handle as usize, 0, buf.as_mut_ptr(), &mut size,
                );
                close_process(&mut pi);
                if ok != 0 && size > 0 {
                    let exe_path = String::from_utf16_lossy(&buf[..size as usize]);
                    // 去掉 \LeagueClient.exe 得到目录
                    if let Some(pos) = exe_path.rfind('\\') {
                        return Some(exe_path[..pos].to_string());
                    }
                    return Some(exe_path);
                }
            }
        }
    }

    // [2] 检查常见安装目录
    let candidates = [
        "C:\\Riot Games\\League of Legends",
        "C:\\Program Files\\Riot Games\\League of Legends",
        "D:\\Riot Games\\League of Legends",
        "E:\\Riot Games\\League of Legends",
    ];
    for c in candidates.iter() {
        let check = format!("{}\\LeagueClient.exe", c);
        let w: Vec<u16> = check.encode_utf16().chain(std::iter::once(0)).collect();
        unsafe {
            extern "system" {
                fn GetFileAttributesW(p: *const u16) -> u32;
            }
            let attr = GetFileAttributesW(w.as_ptr());
            if attr != 0xFFFFFFFF {
                return Some(c.to_string());
            }
        }
    }
    None
}

/// 查找 LOL 游戏进程 (LeagueClient / League of Legends / RiotClientServices)
pub fn find_lol_game_processes() -> Vec<ProcessInfo> {
    let mut out = Vec::new();
    for name in [
        "LeagueClient.exe",
        "League of Legends.exe",
        "RiotClientServices.exe",
    ].iter() {
        if let Some(pi) = find_process(name) {
            out.push(pi);
        }
    }
    out
}
